graph.c

```diff
@@ -188,6 +188,8 @@ edgehash(struct edge *e)
 	} else {
 		e->hash = murmurhash64a(cmd->s, cmd->n);
 	}
+	free(cmd);
+	free(rsp);
```

Comment:
I don't think this works, since it is not necessarily the case that result `edgevar` is a newly allocated string. It might belong to the environment, or the edge.

It is extremely uncommon, but both ninja and samurai allow something like this

```
rule foo
  command = echo abc > $out
build bar: foo
  command = echo 123 > bar
```

When looking up `command`, the edge variable has higher precedence, so `bar` is created with `123` as its content.

An earlier version of samurai used to save expanded rule variables in the edge bindings, but this was problematic since depending on the context, we may need shell-escaped `$in` and `$out`. See #18 and 7ffe36191b91988468e817892905feddb5ee76cc for more context.

Reply:
you are right.